File: leetcode/567.字符串的排列/567.字符串的排列.go
package main

import "fmt"

func checkInclusion(s1 string, s2 string) bool {
	need := map[int32]int{}
	window := map[int32]int{}
	for _, s := range s1 {
		need[s] += 1
		window[s] = 0
	}
	left := 0
	right := 0
	valid := 0
	for right < len(s2){
		c := int32(s2[right])
		right += 1
		if _, ok := need[c]; ok{
			window[c] +=1
			if window[c] == need[c]{
				valid += 1
			}
		}
		for right-left >= len(s1){
			if valid == len(need){
				return true
			}
			d := int32(s2[left])
			left += 1
			if _,ok := need[d];ok{
				if window[d] == need[d]{
					valid -= 1
				}
				window[d] -= 1
			}
		}
	}
	return false
}
func main() {
	s1 := "ab"
	s2 := "eidbaooo"
	result := checkInclusion(s1, s2)
	fmt.Println(result)
}

File: leetcode/567.字符串的排列/567.字符串的排列.py
class Solution:
    # 判断s1是否存在s2排列中
    def checkInclusion(self, s1: str, s2: str) -> bool:
        # 初始化需求字典和窗口字典，用于判断数量
        need = dict()
        window = dict()
        for s in s1:
            need[s] = need.get(s, 0) + 1
            window[s] = 0
        left, right = 0, 0
        valid = 0
        # 扩大窗口，更新window数量，如果是需求字符，更新有效数量
        while right < len(s2):
            c = s2[right]
            right += 1
            if c in need:
                window[c] += 1
                if window[c] == need[c]:
                    valid += 1
            # 缩小窗口
            while (right - left) >= len(s1):
                # 判断是否找到合法子串
                if valid == len(need):
                    return True
                d = s2[left]
                left += 1
                if d in need:
                    if window[d] == need[d]:
                        valid -= 1
                    window[d] -= 1

        return False


if __name__ == '__main__':
    s1 = "ab"
    s2 = "eidbaooo"
    result = Solution().checkInclusion(s1, s2)
    print(result)